llama_agents/services/workflow.py

```diff
@@ -0,0 +1,340 @@
+        """
+        context_hash = state.hash
+        context_str = state.state
+        if not context_str:
+            return workflow
+
+        if hash(context_str + hash_secret) != context_hash:
+            raise ValueError("Context hash does not match. Possible data corruption.")
+
+        # only load context once it's been verified(?)
+        # convert str to bytes
+        context_bytes = base64.b64decode(context_str)
+        context = pickle.loads(context_bytes)
+        workflow._root_context = context
+
+        return workflow
+
+    def dump_workflow_state(self, workflow: Workflow) -> WorkflowState:
+        """Dump the workflow state.
+
+        TODO: This is a bit of a hack.
+        If the context was limited to seriazbale types, this would be easier.
+        """
+        context = workflow._root_context
+        context_bytes = pickle.dumps(context)
+        context_str = base64.b64encode(context_bytes).decode("ascii")
+        context_hash = hash(context_str + hash_secret)
+
+        return WorkflowState(hash=context_hash, state=context_str, run_kwargs={})
+
+    async def processing_loop(self) -> None:
+        """The processing loop for the service.
+
+        TODO: How do we handle any errors that occur during processing?
+        """
+        logger.info("Processing initiated.")
+        while True:
+            if not self.running:
+                await asyncio.sleep(self.step_interval)
+                continue
+
+            async with self.lock:
+                current_calls = [(t, c) for t, c in self._outstanding_calls.items()]
+
+            for task_id, current_call in current_calls:
+                # "fork" the workflow, clear its state
+                self.workflow._queues = {}
+                self.workflow._root_context = {}
+                self.workflow._tasks = set()
+                self.workflow._step_to_context = {}
+                workflow = copy.deepcopy(self.workflow)
+
+                # load the state
+                workflow = self.load_workflow_state(workflow, current_call)
+
+                # run the workflow
+                # TODO: How do we handle streaming? Websockets?
+                result = await workflow.run(**current_call.run_kwargs)
+
+                # dump the state
+                updated_state = self.dump_workflow_state(workflow)
+
+                await self.message_queue.publish(
+                    QueueMessage(
+                        type=CONTROL_PLANE_NAME,
+                        action=ActionTypes.COMPLETED_TASK,
+                        data=TaskResult(
+                            task_id=task_id,
+                            history=[],
+                            result=str(result),
+                            data=updated_state.dict(),
+                        ).model_dump(),
+                    )
+                )
+
+                # clean up
+                async with self.lock:
+                    del self._outstanding_calls[task_id]
```

Comment:
If for whatever reason the `task_id` is not in the dict this will raise a `KeyError`. The preferred approach would be to do:

```suggestion
                    self._outstanding_calls.pop("task_id", None)
```